cpp/benchmarks/groupby/group_max_multithreaded.cpp

```diff
@@ -0,0 +1,96 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <benchmarks/common/generate_input.hpp>
+#include <benchmarks/fixture/benchmark_fixture.hpp>
+
+#include <cudf/detail/utilities/stream_pool.hpp>
+#include <cudf/groupby.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/thread_pool.hpp>
+
+#include <nvbench/nvbench.cuh>
+
+template <typename Type>
+void bench_groupby_max_multithreaded(nvbench::state& state, nvbench::type_list<Type>)
+{
+  auto const cardinality      = static_cast<cudf::size_type>(state.get_int64("cardinality"));
+  auto const num_rows         = static_cast<cudf::size_type>(state.get_int64("num_rows"));
+  auto const null_probability = state.get_float64("null_probability");
+  auto const num_threads      = state.get_int64("num_threads");
+
+  auto const keys = [&] {
+    data_profile const profile =
+      data_profile_builder()
+        .cardinality(cardinality)
+        .no_validity()
+        .distribution(cudf::type_to_id<int32_t>(), distribution_id::UNIFORM, 0, num_rows);
+    return create_random_column(cudf::type_to_id<int32_t>(), row_count{num_rows}, profile);
+  }();
+
+  auto const vals = [&] {
+    auto builder = data_profile_builder().cardinality(0).distribution(
+      cudf::type_to_id<Type>(), distribution_id::UNIFORM, 0, num_rows);
+    if (null_probability > 0) {
+      builder.null_probability(null_probability);
+    } else {
+      builder.no_validity();
+    }
+    return create_random_column(
+      cudf::type_to_id<Type>(), row_count{num_rows}, data_profile{builder});
+  }();
+
+  auto keys_view = keys->view();
+  auto gb_obj    = cudf::groupby::groupby(cudf::table_view({keys_view, keys_view, keys_view}));
+
+  auto streams = cudf::detail::fork_streams(cudf::get_default_stream(), num_threads);
+  cudf::detail::thread_pool threads(num_threads);
+
+  std::vector<std::vector<cudf::groupby::aggregation_request>> requests(num_threads);
+  for (int64_t i = 0; i < num_threads; i++) {
+    requests[i].emplace_back(cudf::groupby::aggregation_request());
+    requests[i][0].values = vals->view();
+    requests[i][0].aggregations.push_back(cudf::make_max_aggregation<cudf::groupby_aggregation>());
+  }
+
+  auto const mem_stats_logger = cudf::memory_stats_logger();
+  state.exec(
+    nvbench::exec_tag::sync | nvbench::exec_tag::timer, [&](nvbench::launch& launch, auto& timer) {
+      auto perform_agg = [&](int64_t index) { gb_obj.aggregate(requests[index], streams[index]); };
+      threads.paused   = true;
+      for (int64_t i = 0; i < num_threads; ++i) {
+        threads.submit(perform_agg, i);
+      }
+      timer.start();
+      threads.paused = false;
+      threads.wait_for_tasks();
```

Comment:
I suspect this contains an sleep delay of 1000microseconds, due to the default-constructed `thread_pool`, and possibly more.

Specifically, each worker thread is spinning doing:
```
    while (running) {
      std::function<void()> task;
      if (!paused && pop_task(task)) {
        task();
        tasks_total--;
      } else {
        sleep_or_yield();
      }
    }
```

So the workers are created on line 60, and immediately `sleep_or_yield`, this calls `sleep_for` the default `sleep_duration` (if it is non-zero) microseconds, or `yield`.

The best-case scenario is that a worker is still sleeping through the `threads.paused = true` assignment on line 73, and the submission of tasks and wakes up after line 78 is executed at which point it can immediately execute. If it wakes before line 78, then even though tasks are available, it is paused and so will go back to sleep.

So I think the pattern wants to be (to match the multi-stream case):

```
timer.start()
for ( int64_t i = 0; i < num_threads; ++i ) {
  threads.submit(perform_agg, i);
}
threads.wait_for_tasks();
join_streams(...);
timer.stop();
```

I note as well that the right way to wake threads in a thread pool is to use condition variables and `notify_all/one` (or semaphores if one only has a single producer/consumer). Indeed, the thread pool implementation that cudf vendors has been updated since we introduced it to use condition variables for notification.

Reply:
Thanks @wence- for the very detailed insights. However, I see no major improvements by applying the changes. 

I re-run benchmarks (see here in [gist](https://gist.github.com/srinivasyadav18/12de3c74dda5312a5a052ff6ac204b9d#0-tesla-t4-2)). From my obeservations, there is always a consistent **< 1ms** overhead due to threads, which becomes neglibable by adding more work (number of aggregations to perform) to each thread.

Reply:
Yeah, I think this is an inevitability of the sleep_for call. What happens if you set sleep_duration to zero on the thread pool you create?

Reply:
There is very slight improvement on low aggregation count (see here in [gist](https://gist.github.com/srinivasyadav18/1ff564a602f9d494af3d117f89f2111c)).

Reply:
> There is very slight improvement on low aggregation count (see here in gist).

How about the sleep delay? Are they still present when profiling with `nsys`?

Reply:
I made an error (in defaulting constructor's argument for `sleep_duration` in `thread_pool` class) in previous benchmarks results. Please ignore them.

Here are new results: [see_in_gist ](https://gist.github.com/srinivasyadav18/fb8bf574f527685ddbe5bc9cf54fb8e1)

This show's **no** threadoverheads both in benchmarks and profiling. So, making the sleep time as 0 helped the performance to be same **groupby_max** scalar version.

Profiling with sleep time = 1000ms
![image (1)](https://github.com/rapidsai/cudf/assets/43375352/52fe5d5d-c6e7-4a7b-9fd8-dabe6c1cd5bf)

Profling with sleep time = 0



![image (2)](https://github.com/rapidsai/cudf/assets/43375352/86bf589c-9720-4b30-9aa2-fce3ba0e838c)

Reply:
@wence- When making `sleep_time` 0, the `thread_pool` will always put the `thread to yield`, which might have different side-effects for other use cases which could effect performance. For this use case, its giving us performance similar to scalar case (i.e with no overheads), but I am not sure how this might impact in a case when threads have more amount of work to do.

Reply:
Thanks, this looks better! The right thing to do is to (as noted) use condition variables, rather than busy-wait with OS hint. This is what the updated version of the vendored thread_pool implementation does. I will open a PR to update the version.